### DISCUSSION TITLE
return additional options/parent info on event changes

### DISCUSSION THREAD

**Discussion by avc-choy** [cocopon/tweakpane#492] [Ideas]
currently when capturing events globally - ie. `pane.on('change', (ev) => {}`, the only really useful members are `presetKey` and `value`. 

however, it would be extremely useful to access the name/info of the parent controllers such as the folder and tabs as well. or, even further, the full dictionary passed in when adding the UI elements via `addInput`/ `addButton`. 

this way, we could grab additional data such as `min` / `max` values of elements - or custom keys we pipe in to do additional logic handling. 

as far as I could tell - the included `target` member of the event handler doesn't provide any of this information either...

am I missing an existing way to do this? would be great to get this feature added if not.

**Comment by cocopon** [comment]
In the next major version v4, `target` (`BladeApi`, `InputBindingApi`, or `BindingApi`) will have more info:

- Key of the bound value (former `presetKey`)
- Element
- Bound value
- Blade label
- Generic `tag` string
- Other API-specific values
  - e.g. `min`, `max` for the slider, `options` for the list, ...

Currently I don't have a plan to add a parent or initial options. See #396 for ongoing specs.

**Reply by avc-choy** [reply]
that's great, thanks.

will you be providing access to all the key/vals passed through when creating inputs - in addition to the API-specific values (ie. `min`, `max`, `options`)? this would allow us to include custom keys and handle additional logic quite nicely.

for the generic `tag` string - will this be something we can supply when creating the input? or, will it be supplied automatically?

**Reply by cocopon** [reply]
> will you be providing access to all the key/vals passed through when creating inputs

Not all values. It's not easy to make a property mutable. For example, here is the PR that makes `min`, `max` and `options` mutable:
https://github.com/cocopon/tweakpane/pull/449/files

More test cases means more complexity.

> will this be something we can supply when creating the input?

Right. You'll be able to use it in v4 as follows:

```js
const b = pane.addBinding(PARAMS, 'foo', {
  tag: 'something',
});

console.log(b.tag);  // 'something'
```

**Reply by avc-choy** [reply]
understood - thanks! looking forward to v4...
